frontend/components/generateProof.tsx

```diff
@@ -0,0 +1,199 @@
+  const [loadingText, setLoadingText] = React.useState<string | undefined>(
+    undefined
+  );
+  const [successProofGen, setSuccessProofGen] = React.useState<
+    boolean | undefined
+  >(undefined);
+
+  const [groupId, setGroupId] = React.useState<number>(1);
+
+  const { data, error, isLoading, signMessage } = useSignMessage({
+    async onSuccess(data, variables) {
+      // Verify signature when sign message succeeds
+      const { v, r, s } = ethers.utils.splitSignature(data);
+      const isYOdd = (BigInt(v) - BigInt(27)) % BigInt(2);
+      const rPoint = ec.keyFromPublic(
+        ec.curve.pointFromX(new BN(r.substring(2), 16), isYOdd).encode("hex"),
+        "hex"
+      );
+      // Get the group element: -(m * r^−1 * G)
+      const rInv = new BN(r.substring(2), 16).invm(SECP256K1_N);
+
+      // w = -(r^-1 * msg)
+      const w = rInv
+        .mul(
+          new BN(ethers.utils.hashMessage(variables.message).substring(2), 16)
+        )
+        .neg()
+        .umod(SECP256K1_N);
+      // U = -(w * G) = -(r^-1 * msg * G)
+      const U = ec.curve.g.mul(w);
+
+      // T = r^-1 * R
+      const T = rPoint.getPublic().mul(rInv);
+
+      const TPreComputes = await getPointPreComputes(T.encode("hex"));
+      const signatureArtifacts: SignaturePostProcessingContents = {
+        TPreComputes,
+        s: splitToRegisters(s.substring(2)) as bigint[],
+        U: [
+          splitToRegisters(U.x) as bigint[],
+          splitToRegisters(U.y) as bigint[],
+        ],
+      };
+      await generateProof(signatureArtifacts);
+    },
+  });
+
+  const generateProof = React.useCallback(
+    async (artifacts: SignaturePostProcessingContents) => {
+      if (!merkleTreeProofData.current) {
+        throw new Error("Missing merkle tree data");
+      }
+      //TODO: add loading state or progress bar first time it downloads zkey
+      await downloadZKey();
+
+      const proofInputs = { ...artifacts, ...merkleTreeProofData.current };
+      const zkeyDb = await localforage.getItem("setMembership_final.zkey");
+
+      if (!zkeyDb) {
+        throw new Error("zkey was not found in the database");
+        return;
+      }
+
+      // @ts-ignore
+      const zkeyRawData = new Uint8Array(zkeyDb);
+
+      const zkeyFastFile = { type: "mem", data: zkeyRawData };
+      const worker = new Worker("./worker.js");
+      worker.postMessage([proofInputs, zkeyFastFile]);
+      worker.onmessage = async function (e) {
+        const { proof, publicSignals } = e.data;
+        console.log("PROOF SUCCESSFULLY GENERATED: ", proof, publicSignals);
+        // TODO: post to IPFS or store in our db
+        setSuccessProofGen(true);
+        // TODO: add toast showing success
+        setLoadingText(undefined);
+      };
+    },
+    []
+  );
+
+  const prepareProof = React.useCallback(async () => {
+    try {
+      const merkleTreeData = (
+        await axios.get("/api/getPropGroup", {
+          params: {
+            userAddr: address,
```

Comment:
we definitely don't want the user to dox the address they're proving for to the backend. probably makes sense to get all trees for prop or group and filter here based on address. i had a route like this here: https://github.com/personaelabs/nouns150/blob/lsankar/address-to-groups/frontend/pages/api/getPropTrees.ts#L21

although probably makes sense to change the route you have here to be by group/prop